### ISSUE TITLE
Add .name to macro context

### ISSUE THREAD

**Issue by disnet** [mozilla/sweet.js#538]
Need a way of retrieving the invoking macro's name syntax object:

```js
syntax m = ctx => {
  ctx.name(); // m
}
m 42;
```

**Comment by bopjesvla** [comment]
I think you mentioned in #529 that unlike `#'m'`, using `ctx.name()` would make `m` recursive (edited for this example):

```
Definitely, that was in my mind but didn’t make it into code for the MVP. Trivial to add but not critical until we have syntax creation functions and the need to perform hygiene shenanigans.

In this case though, you don’t want ctx.name() since that would just make it a recursive definition. You want the previous binding via #'m'.
```

Wouldn't this make `syntaxrec` superfluous?

**Comment by disnet** [comment]
`.name` gets the syntax object at the call-site (along with its associated scope) while `syntaxrec` gets the syntax object at the definition site. The call-site syntax object is often used in Scheme/Racket for hygiene bending:

```js
import { makeIdentifier } from 'syntax' for syntax; // or something similar

let x = 'outer';
syntax m = ctx => {
    return #`${makeIdentifier('x', ctx.name()}`;  // makes a new identifier bound in the scope of ctx.name()
}

function foo(x) {
  let x = 'inner';
  m; // expands to the inner x binding
}
```

**Comment by bopjesvla** [comment]
Ah, that makes sense, thanks!

**Comment by disnet** [comment]
Closed with #544
